src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -766,5 +767,45 @@ private async Task<bool> CompleteOutstandingActiveRequests(string targetName)
 
             return false;
         }
+
+        private void ComputeAfterTargetFailures(string[] targetNames)
+        {
+            // Cache intermediate results so we don't visit a target more than once.
+            Dictionary<string, bool> resultsWithAfterTargets = new Dictionary<string, bool>();
+
+            foreach (string targetName in targetNames)
+            {
+                if (_buildResult.ResultsByTarget.ContainsKey(targetName))
+                {
+                    // Queue of targets waiting to be processed, seeded with the specific target for which we're computing AfterTargetsHaveFailed.
+                    Queue<string> targetsToCheckForAfterTargets = new Queue<string>();
+                    targetsToCheckForAfterTargets.Enqueue(targetName);
+
+                    while (targetsToCheckForAfterTargets.Count > 0)
+                    {
+                        string targetToCheck = targetsToCheckForAfterTargets.Dequeue();
+                        IList<TargetSpecification> targetsWhichRunAfter = _requestEntry.RequestConfiguration.Project.GetTargetsWhichRunAfter(targetToCheck);
+
+                        foreach (TargetSpecification afterTarget in targetsWhichRunAfter)
+                        {
+                            if (!resultsWithAfterTargets.TryGetValue(afterTarget.TargetName, out bool afterTargetFailed))
```

Comment:
minor: should we log anything if `TryGetValue` returns true?

Reply:
No, I don't believe that this is a log-worthy event. We're simply caching the results in a local dictionary to avoid getting into the O(n^2) run-time territory. A cache hit/miss is pure implementation and not interesting outside of this method.